{"metadata": {"kernelspec": {"name": "conda-root-py", "display_name": "Python [conda env:root] *", "language": "python"}, "language_info": {"name": "python", "version": "3.7.3", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "## `00_fetchreads`: This notebook shows how we fetch reads that may belong to the MHC region of the HG002 Genomes\n\nWe use Peregrine command `shmr-map` to compare the CCS reads for phasing and assembling the HG002 MHC region. The 'shmr-map' tools using SHIMMER (https://www.biorxiv.org/content/10.1101/705616v1) to map the reads to the GHRh37 MHC and the (un-phased) genome assembly of HG002 MHC region from an assembly (s3://human-pangenomics/HPRC/HG002_Assessment/assemblies/JC_20k_15k_asm/asm.fa.gz, contig 000028F:1700756-6708745). The two MHC sequeneces are combined in the file `MHC_all.fa` for recuriting.", "metadata": {}}, {"cell_type": "code", "source": "%%bash\ndx download -f /jc_notebook/MHC_37.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "raw", "source": "aws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/15kb/m64012_190920_173625.Q20.fastq hg002_reads/\naws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/15kb/m64012_190921_234837.Q20.fastq hg002_reads/\naws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/20kb/m64011_190830_220126.Q20.fastq hg002_reads/\naws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/20kb/m64011_190901_095311.Q20.fastq hg002_reads/\n", "metadata": {}}, {"cell_type": "code", "source": "%%capture\n%%bash\nconda install -y matplotlib bokeh", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx download /20200316_asm_for_revision/data/MHC_37.fa\ndx download /20200316_asm_for_revision/data/MHC_all.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "!cat MHC_37.fa MHC_asm.fa >  /home/dnanexus/MHC_all.fa\n!ls -l  /home/dnanexus/MHC_all.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "MHC on GRC37: chr6:28,477,797-33,448,354", "metadata": {}}, {"cell_type": "code", "source": "%%writefile /home/dnanexus/build_read_index_and_map.sh\ncd /home/dnanexus/\n\necho  MHC_all.fa > ref.lst\nshmr_mkseqdb -d ref.lst -p ref\nshmr_index -p ref -o ref-shmr -r 3\n\nfind /home/dnanexus/hg002_reads -name \"*.fastq\" > reads.lst\nshmr_mkseqdb -d reads.lst -p reads \n\nmkdir -p read_index\nfor i in `seq 1 24`; do echo shmr_index -p reads -o read-shmr -r 3 -t 24 -c $i; done | \\\nparallel -j 12\nfor i in `seq 1 24`; do echo \"shmr_map -r ref -m ./ref-shmr-L2 -p reads -l read-shmr-L2 -t 24 -c $i > map.$i\"; done | \\\nparallel -j 6", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "raw", "source": "%%bash\nbash build_read_index_and_map.sh", "metadata": {}}, {"cell_type": "code", "source": "%%bash\ncd /home/dnanexus/\ncat map.* | sort -g -k1g -k2g --parallel 24 -T /tmp  > map.sorted\ncat map.sorted | awk '{print $4}' | \\\nsort -g  | uniq -c  | awk '$1 > 2 {print $2}' > mapped.readids", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "!wc /home/dnanexus/mapped.readids", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "all_readids = set([int(_) for _ in open(\"/home/dnanexus/mapped.readids\").read().split()])", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "read_sdb = SequenceDatabase(\"/home/dnanexus/reads.idx\", \"/home/dnanexus/reads.seqdb\")", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "with open(\"/home/dnanexus/mapped_reads.fa\", \"w\") as f:\n    for rid in all_readids:\n        seq_name = read_sdb.index_data[rid].rname\n        seq = read_sdb.get_subseq_by_rid(rid)\n        print(f\">{seq_name}\", file=f)\n        print(seq.decode(), file=f)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "!grep -c \">\" /home/dnanexus/mapped_reads.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}]}