{"metadata": {"kernelspec": {"name": "conda-root-py", "display_name": "Python [conda env:root] *", "language": "python"}, "language_info": {"name": "python", "version": "3.7.3", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "## `01_get_phased_reads`: Generating Phased Read Sets\n\nThis notebook shows the process of performing the haplotype phasing with 10x, Oxford Nanopore Reads and PacBio CCS reads. We also perform a couple different experiments of different combinations of the data sets for phasing.", "metadata": {}}, {"cell_type": "markdown", "source": "## Getting GIAB data and selectings regions covering MHC region", "metadata": {}}, {"cell_type": "markdown", "source": "The data was downloaded from the following links:\n\nPacBio CCS reads from human pangenomics project \n - aws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/15kb/m64012_190920_173625.Q20.fastq 15kb/\n - aws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/15kb/m64012_190921_234837.Q20.fastq 15kb/\n - aws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/20kb/m64011_190830_220126.Q20.fastq 15kb/\n - aws s3 cp s3://human-pangenomics/HG002/hpp_HG002_NA24385_son_v1/PacBio_HiFi/20kb/m64011_190901_095311.Q20.fastq 15kb/\n\nONT \"ultralong\" \n - http://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/HG002_NA24385_son/Ultralong_OxfordNanopore/guppy-V3.2.4_2020-01-22/HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam \n - ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/HG002_NA24385_son/Ultralong_OxfordNanopore/guppy-V3.2.4_2020-01-22/HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam.bai    \n", "metadata": {}}, {"cell_type": "markdown", "source": "### for running this notebook, we need to setup a conda env and IPython kernel using a unix shell console. \n```\nconda create -n whatshap\nconda activate whatshap\nconda install -y minimap2 samtools bcftools pysam bamtools whatshap -c bioconda\n\napt-get install samtools bcftools bamtools bzip2 tabix\n#pip install git+https://bitbucket.org/whatshap/whatshap@split\n\n```\nWe need to reconnect the new IPython kernel to make this work", "metadata": {}}, {"cell_type": "code", "source": "%%bash\ndx download -f /GIAB_BAM_files/HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam\ndx download -f /GIAB_BAM_files/HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam.bai", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\napt-get install -y pigz\n#pip install git+https://bitbucket.org/whatshap/whatshap@split", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "For PacBio CCS Sequel 15Kb, Sequel II 11Kb, and ONT \"ultra-long\" the reads covering the MHC were selected using the following commands (with bam file name updated for each data) for haplotype 1 and haplotype 2:\n\nThis script was run with a directory structure that includes subdirectories of \"reads\", cram\", \"ref\", \"vcf\", \"whatshap\"", "metadata": {}}, {"cell_type": "code", "source": "%%writefile filter_script.json\n{ \"filters\" : [ { \"id\" : \"inHP1\", \"tag\" : \"HP:1\" }, { \"id\" : \"inHP2\", \"tag\" : \"HP:2\" } ], \"rule\" : \"!(inHP1 | inHP2)\" }", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "filter_script.json:\n{\n    \"filters\" : [\n    { \"id\" : \"inHP1\", \"tag\" : \"HP:1\" },\n    { \"id\" : \"inHP2\", \"tag\" : \"HP:2\" }\n    ],\n    \"rule\" : \"!(inHP1 | inHP2)\"\n}", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n\nsamtools view -bh HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam 6:28477797-33448354 | \\\n    bamtools filter -in stdin -tag HP:1 | \\\n        samtools bam2fq - | \\\n        bgzip -c > HG002.ultra-long-ont_hs37d5_phased_reheader.HP1.fastq.gz\n\nsamtools view -bh HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam 6:28477797-33448354 | \\\n    bamtools filter -in stdin -tag HP:2 | \\\n        samtools bam2fq - | \\\n        bgzip -c > HG002.ultra-long-ont_hs37d5_phased_reheader.HP2.fastq.gz\n\nsamtools view -bh HG002_hs37d5_ONT-UL_GIAB_20200122.phased.bam 6:28477797-33448354 | \\\n    bamtools filter -in stdin -script filter_script.json | \\\n    samtools bam2fq - | \\\n    bgzip -c > HG002.ultra-long-ont_hs37d5_phased_reheader.MHConly.UnPartitioned.fastq.gz", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "## Generate haplotype binned reads with various tools within the WhatsHAP toolset.", "metadata": {}}, {"cell_type": "markdown", "source": "Getting the input files", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n#wget --quiet ftp://ftp-trace.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/analysis/10XGenomics_ChromiumGenome_LongRanger2.2_Supernova2.0.1_04122018/GRCh37/NA24385_300G/NA24385.GRCh37.phased_variants.vcf.gz\n#wget --quiet https://storage.cloud.google.com/genomics-public-data/references/hs37d5/hs37d5.fa.gz\n\ndx download -f /20200316_asm_for_revision/data/NA24385.GRCh37.phased_variants.vcf.gz\ndx download -f /20200316_asm_for_revision/data/hs37d5.fa.gz\n", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nmkdir -p ref\nmv hs37d5.fa.gz ref", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx download -f /20200316_asm_for_revision/data/mapped_reads.fa\ngrep -c \">\" mapped_reads.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Merging FASTQs again\nconvert mapped_reads.fa to reads/HG002.Sequel.15kb.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.fastq.gz", "metadata": {}}, {"cell_type": "code", "source": "%%bash\napt-get install pigz\nmkdir -p reads\nzcat HG002.ultra-long-ont_hs37d5_phased_reheader.*.fastq.gz | pigz > reads/HG002_ultra-long-ont_hs37d5_phased_reheader.MHConly.fastq.gz\ncat mapped_reads.fa | pigz > reads/HG002.Sequel.15kb.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.fastq.gz\n#cat HG002.Sequel.15kb.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.fastq |  pigz > reads/HG002.Sequel.15kb.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.fastq.gz\n#zcat HG002.SequelII.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.*.fastq.gz | pigz >> reads/HG002.Sequel.15kb.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.fastq.gz", "metadata": {"scrolled": true, "trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Mapping CCS reads", "metadata": {}}, {"cell_type": "code", "source": "%%bash\nmkdir -p cram\n\n/opt/conda/envs/whatshap/bin/minimap2 -t 32 -R '@RG\\tID:1\\tSM:HG002' -ax asm20 ref/hs37d5.fa.gz \\\nreads/HG002.Sequel.15kb.pbmm2.hs37d5.whatshap.haplotag.RTG.10x.trio.MHConly.fastq.gz | \\\nsamtools view -bS > cram/HG002.PacBio.15kbCCS.bam", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nsamtools sort -@ 32 -o cram/HG002.PacBio.15kbCCS.sorted.bam cram/HG002.PacBio.15kbCCS.bam\nsamtools index cram/HG002.PacBio.15kbCCS.sorted.bam", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Mapping ONT reads", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n/opt/conda/envs/whatshap/bin/minimap2 -t 32 -R '@RG\\tID:1\\tSM:HG002' -ax map-ont ref/hs37d5.fa.gz \\\nreads/HG002_ultra-long-ont_hs37d5_phased_reheader.MHConly.fastq.gz | \\\nsamtools view -bS - > cram/HG002.ultra-long-ont_hs37d5_phased_reheader.bam", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nsamtools sort  -@ 32  -o cram/HG002.ultra-long-ont_hs37d5_phased_reheader.sorted.bam cram/HG002.ultra-long-ont_hs37d5_phased_reheader.bam \nsamtools index cram/HG002.ultra-long-ont_hs37d5_phased_reheader.sorted.bam", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx download -f /HG002_reanalysis/pacbio-15kb-hapsort-wgs.vcf.gz\nmkdir -p vcf/deepvariant/\nmv pacbio-15kb-hapsort-wgs.vcf.gz vcf/deepvariant/", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ntabix -p vcf vcf/deepvariant/pacbio-15kb-hapsort-wgs.vcf.gz", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Select het SNVs from DeepVariant call set. The goal is to use this as a starting point to construct a super confident set of heterozygous SNVs that we can use to partition reads", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n/opt/conda/envs/whatshap/bin/bcftools view -g het -v snps vcf/deepvariant/pacbio-15kb-hapsort-wgs.vcf.gz 6:28477797-33448354 > vcf/deepvariant/mhc.het.snps.vcf", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Re-genotype these SNV from ONT reads", "metadata": {}}, {"cell_type": "code", "source": "%%bash\nzcat ref/hs37d5.fa.gz > ref/hs37d5.fa\nsamtools faidx ref/hs37d5.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nmkdir -p whatshap/genotype/\n/opt/conda/envs/whatshap/bin/whatshap genotype --ignore-read-groups --chromosome 6 \\\n--reference ref/hs37d5.fa \\\n-o whatshap/genotype/HG002.ultra-long-ont.vcf \\\nvcf/deepvariant/mhc.het.snps.vcf \\\ncram/HG002.ultra-long-ont_hs37d5_phased_reheader.sorted.bam > whatshap/genotype/HG002.ultra-long-ont.log 2>&1", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Re-genotype these SNV from CCS reads", "metadata": {}}, {"cell_type": "code", "source": "%%bash\n/opt/conda/envs/whatshap/bin/whatshap genotype --ignore-read-groups --chromosome 6 \\\n--reference ref/hs37d5.fa \\\n-o whatshap/genotype/HG002.PacBio.15kbCCS.vcf \\\nvcf/deepvariant/mhc.het.snps.vcf \\\ncram/HG002.PacBio.15kbCCS.sorted.bam > whatshap/genotype/HG002.PacBio.15kbCCS.log 2>&1", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Find the intersection, i.e. only retain those that are het accoring to DV/CCS, WhatHap/ONT, and WhatsHap/CCS", "metadata": {}}, {"cell_type": "code", "source": "%%bash\nbcftools view -g het whatshap/genotype/HG002.ultra-long-ont.vcf > whatshap/genotype/HG002.ultra-long-ont.hetonly.vcf\nbgzip -f whatshap/genotype/HG002.ultra-long-ont.hetonly.vcf\nrm HG002.ultra-long-ont.hetonly.vcf.gz.tbi\ntabix -p vcf -f whatshap/genotype/HG002.ultra-long-ont.hetonly.vcf.gz\n\nbcftools view -g het whatshap/genotype/HG002.PacBio.15kbCCS.vcf > whatshap/genotype/HG002.PacBio.15kbCCS.hetonly.vcf\nbgzip -f whatshap/genotype/HG002.PacBio.15kbCCS.hetonly.vcf\nrm HG002.PacBio.15kbCCS.hetonly.vcf.gz.tbi\ntabix -p vcf -f whatshap/genotype/HG002.PacBio.15kbCCS.hetonly.vcf.gz\n\nmkdir -p whatshap/genotype-intersect\nbcftools isec -p whatshap/genotype-intersect whatshap/genotype/HG002.ultra-long-ont.hetonly.vcf.gz whatshap/genotype/HG002.PacBio.15kbCCS.hetonly.vcf.gz \ncp whatshap/genotype-intersect/0003.vcf whatshap/confident-hets.vcf", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Extract 10x phased blocks produced by LongRanger ", "metadata": {}}, {"cell_type": "code", "source": "%%bash\nmkdir -p 10x\ncp NA24385.GRCh37.phased_variants.vcf.gz 10x/\ncd 10x\nrm NA24385.GRCh37.phased_variants.vcf.gz.tbi\ntabix -p vcf -f NA24385.GRCh37.phased_variants.vcf.gz\nbcftools view NA24385.GRCh37.phased_variants.vcf.gz chr6:28477797-33448354 | \\\n    awk 'BEGIN {OFS=\"\\t\"} $1 == \"chr6\" {$1=\"6\"} $1==\"#CHROM\" {$10=\"HG002\"} {print}'| \\\n    bgzip > NA24385.GRCh37.phased_variants.mhc.vcf.gz", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "markdown", "source": "Now run phasing from multiple combinations of data sources", "metadata": {}}, {"cell_type": "code", "source": "%%bash\nrm -rf whatshap/phase/\nmkdir -p whatshap/phase/\n\n/opt/conda/envs/whatshap/bin/whatshap phase --ignore-read-groups --chromosome 6 \\\n--reference ref/hs37d5.fa \\\n-o whatshap/phase/HG002.PacBio.15kbCCS.vcf \\\nwhatshap/confident-hets.vcf \\\ncram/HG002.PacBio.15kbCCS.sorted.bam > whatshap/phase/HG002.PacBio.15kbCCS.log 2>&1\n\nbgzip -f whatshap/phase/HG002.PacBio.15kbCCS.vcf\ntabix -p vcf -f whatshap/phase/HG002.PacBio.15kbCCS.vcf.gz\n\n/opt/conda/envs/whatshap/bin/whatshap phase --ignore-read-groups --chromosome 6 \\\n--reference ref/hs37d5.fa -o whatshap/phase/HG002.ultra-long-ont.vcf \\\nwhatshap/confident-hets.vcf \\\ncram/HG002.ultra-long-ont_hs37d5_phased_reheader.sorted.bam > whatshap/phase/HG002.ultra-long-ont.log 2>&1\n\nbgzip -f whatshap/phase/HG002.ultra-long-ont.vcf\ntabix -p vcf -f  whatshap/phase/HG002.ultra-long-ont.vcf.gz\n\n/opt/conda/envs/whatshap/bin/whatshap phase --chromosome 6 \\\n--reference ref/hs37d5.fa \\\n-o whatshap/phase/HG002.ultra-long-ont+10x.vcf \\\nwhatshap/confident-hets.vcf \\\ncram/HG002.ultra-long-ont_hs37d5_phased_reheader.sorted.bam \\\n10x/NA24385.GRCh37.phased_variants.mhc.vcf.gz > whatshap/phase/HG002.ultra-long-ont+10x.log  2>&1\n\nbgzip -f whatshap/phase/HG002.ultra-long-ont+10x.vcf\ntabix -p vcf -f  whatshap/phase/HG002.ultra-long-ont+10x.vcf.gz\n\n/opt/conda/envs/whatshap/bin/whatshap phase --ignore-read-groups --chromosome 6 \\\n--reference ref/hs37d5.fa \\\n-o whatshap/phase/HG002.PacBio+10x.vcf \\\nwhatshap/confident-hets.vcf cram/HG002.PacBio.15kbCCS.sorted.bam \\\n10x/NA24385.GRCh37.phased_variants.mhc.vcf.gz > whatshap/phase/HG002.PacBio+10x.log 2>&1\n\nbgzip -f whatshap/phase/HG002.PacBio+10x.vcf\ntabix -p vcf -f whatshap/phase/HG002.PacBio+10x.vcf.gz\n\n", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\n/opt/conda/envs/whatshap/bin/whatshap phase --ignore-read-groups --chromosome 6 \\\n--reference ref/hs37d5.fa \\\n-o whatshap/phase/HG002.PacBio+ultra-long-ont.vcf \\\nwhatshap/confident-hets.vcf cram/HG002.PacBio.15kbCCS.sorted.bam \\\nwhatshap/phase/HG002.ultra-long-ont.vcf.gz > whatshap/phase/HG002.PacBio+ultra-long-ont.log 2>&1", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\n/opt/conda/envs/whatshap/bin/whatshap haplotag --ignore-read-groups \\\n--reference ref/hs37d5.fa \\\n-o cram/HG002.PacBio.15kbCCS.sorted.tagged-newsplit.bam \\\nwhatshap/phase/HG002.ultra-long-ont+10x.vcf.gz \\\ncram/HG002.PacBio.15kbCCS.sorted.bam", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nsamtools view cram/HG002.PacBio.15kbCCS.sorted.tagged-newsplit.bam | grep HP:i:1 | awk '{print $1}' | sort -u  > H1_reads &\nsamtools view cram/HG002.PacBio.15kbCCS.sorted.tagged-newsplit.bam | grep HP:i:2 | awk '{print $1}' | sort -u  > H2_reads &\nsamtools view cram/HG002.PacBio.15kbCCS.sorted.tagged-newsplit.bam | grep -v HP:i:2 | grep -v HP:i:1 | awk '{print $1}' | sort -u > tmp_reads &", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "phased_reads = open(\"H1_reads\").read().split(\"\\n\")\nphased_reads.extend(open(\"H2_reads\").read().split(\"\\n\"))\ntmp_reads = open(\"tmp_reads\").read().split(\"\\n\")\ntmp_reads = set(tmp_reads)\nphased_reads = set(phased_reads)\nunphased_reads = tmp_reads - phased_reads\nwith open(\"unphased_reads\", \"w\") as f:\n    for r in unphased_reads:\n        print(r, file=f)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}