{"metadata": {"language_info": {"name": "python", "version": "3.7.3", "mimetype": "text/x-python", "codemirror_mode": {"name": "ipython", "version": 3}, "pygments_lexer": "ipython3", "nbconvert_exporter": "python", "file_extension": ".py"}, "kernelspec": {"name": "python3", "display_name": "Python 3", "language": "python"}}, "nbformat_minor": 4, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "## `03_get_variant_cluster`: Get variants called from the reads mapped back to the assembled contigs\n\nThe assembled contigs should be consistent with the reads that used to generate them. We mapped the reads back to the contigs and check if there are clusters of variants to find potenting issues of assembly or consensus errors.", "metadata": {}}, {"cell_type": "code", "source": "%%bash\ndx cd /20200316_asm_for_revision/\ndx download -f H1_pa_phase_ctg.fa H2_pa_phase_ctg.fa\ndx download -f data/H1.fa data/H2.fa data/unphased.fa\ndx download -f data/MHC_37.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%capture\n%%bash\n. /opt/conda/bin/activate\nconda install -c bioconda -y freebayes=1.3.1\nconda install -c bioconda -y minimap2 samtools pbmm2 pbsv", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\n. /opt/conda/bin/activate\ncat H1.fa unphased.fa > H1_and_unphased.fa\n\npbmm2 align H1_pa_phase_ctg.fa H1_and_unphased.fa --sort --preset CCS --sample sample1 --rg '@RG\\tID:movie1' > H1_to_H1asm.bam \nsamtools index H1_to_H1asm.bam\ncat H2.fa unphased.fa > H2_and_unphased.fa\n\npbmm2 align H2_pa_phase_ctg.fa H2_and_unphased.fa --sort --preset CCS --sample sample1 --rg '@RG\\tID:movie1' > H2_to_H2asm.bam\nsamtools index H2_to_H2asm.bam", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nsamtools faidx -f H1_pa_phase_ctg.fa\nsamtools faidx -f H2_pa_phase_ctg.fa", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash \n# this takes about one hours to run\nfreebayes-parallel <(fasta_generate_regions.py H1_pa_phase_ctg.fa.fai 500000) 8 \\\n   -G 10 -F 0.3  -f H1_pa_phase_ctg.fa H1_to_H1asm.bam > var_H1_2.vcf &\n\nfreebayes-parallel <(fasta_generate_regions.py H2_pa_phase_ctg.fa.fai 500000) 8 \\\n   -G 10 -F 0.3  -f H2_pa_phase_ctg.fa H2_to_H2asm.bam > var_H2_2.vcf &\n", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\nbgzip -f var_H1_2.vcf\nbgzip -f var_H2_2.vcf\n", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ntabix -p vcf var_H1_2.vcf.gz\ntabix -p vcf var_H2_2.vcf.gz", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\npbsv discover H1_to_H1asm.bam H1_to_H1asm.svsig.gz\npbsv discover H2_to_H2asm.bam H2_to_H2asm.svsig.gz\npbsv call H1_pa_phase_ctg.fa H1_to_H1asm.svsig.gz H1_to_H1asm.sv.vcf\npbsv call H2_pa_phase_ctg.fa H2_to_H2asm.svsig.gz H2_to_H2asm.sv.vcf\nbgzip -f H1_to_H1asm.sv.vcf \nbgzip -f H2_to_H2asm.sv.vcf", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx cd /20200316_asm_for_revision\ndx upload var_H1_2.vcf.gz\ndx upload var_H2_2.vcf.gz\ndx upload H2_to_H2asm.sv.vcf.gz\ndx upload H2_to_H2asm.sv.vcf.gz ", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "import gzip\n\nf = gzip.open(\"var_H1_2.vcf.gz\")\nv_pos = []\nfor row in f:\n    if row[0] == ord(b'#'):\n        continue\n    row = row.strip().split()\n    if row[0] != b'000000F':\n        continue\n    if len(row[3]) != len(row[4]): # only count for SNP\n        continue\n    position = int(row[1])\n    v_pos.append(position)\n\nf = gzip.open(\"H1_to_H1asm.sv.vcf.gz\") # from pb_sv call \nfor row in f:\n    if row[0] == ord(b'#'):\n        continue\n    row = row.strip().split()\n    if row[0] != b'000000F':\n        continue\n    position = int(row[1])\n    k, v = row[-2:]\n    k = k.split(b':')\n    v = v.split(b':')\n    d = dict(zip(k,v))\n    if b'AD' not in d:\n        continue\n    AD = [int(_) for _ in d[b'AD'].split(b',')]\n    if min(AD)/sum(AD) < 0.25:\n        continue\n    pos = int(row[1])\n    v_pos.append(position)\n\nv_pos.sort()\nSNP_cluster=[]\ncluster_d = 10000\nfor position in v_pos:\n    if len(SNP_cluster) == 0 or position - SNP_cluster[-1][-1] > cluster_d:\n        SNP_cluster.append([position])\n    else:\n        SNP_cluster[-1].append(position)\nf = open(\"H1_exclude.bed\",\"w\")\nfor c in SNP_cluster:\n    #print(\"H1\", len(c), '000000F:{}-{}'.format(min(c)-50, max(c)+50), max(c)-min(c)+100)\n    print('000000F\\t{}\\t{}'.format(min(c)-50, max(c)+50), file=f)\nf.close()\nfor c in SNP_cluster:\n    print((min(c)-50, max(c)+50), max(c)-min(c)+100)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\napt-get install zlib1g-dev\ngit clone https://github.com/lh3/seqtk.git;\ncd seqtk; make", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\n./seqtk/seqtk subseq H1_pa_phase_ctg.fa H1_exclude.bed > out.fa\ngrep \">\" out.fa\nminimap2 -xasm5 MHC_37.fa  out.fa > out.paf\ncat out.paf | awk '$12 >0 {print 6\" \"$8+28477798\" \"$9+28477798\" H1:\"$1\":\"$12}' | sort -k2 -g > H1_exclude_GRCH37.bed", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "!cat H1_exclude_GRCH37.bed", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "f = gzip.open(\"var_H2_2.vcf.gz\")\nv_pos = []\nfor row in f:\n    if row[0] == ord(b'#'):\n        continue\n    row = row.strip().split()\n    if row[0] != b'000000F':\n        continue\n    if len(row[3]) != len(row[4]):\n        continue\n    position = int(row[1])\n    v_pos.append(position)\n\nf = gzip.open(\"H2_to_H2asm.sv.vcf.gz\")\nfor row in f:\n    if row[0] == ord(b'#'):\n        continue\n    row = row.strip().split()\n    if row[0] != b'000000F':\n        continue\n    position = int(row[1])\n    k, v = row[-2:]\n    k = k.split(b':')\n    v = v.split(b':')\n    d = dict(zip(k,v))\n    if b'AD' not in d:\n        continue\n    AD = [int(_) for _ in d[b'AD'].split(b',')]\n    if min(AD)/sum(AD) < 0.25:\n        continue\n    pos = int(row[1])\n    v_pos.append(position)\n\nv_pos.sort()\nSNP_cluster=[]\ncluster_d = 10000\nfor position in v_pos:\n    if len(SNP_cluster) == 0 or position - SNP_cluster[-1][-1] > cluster_d:\n        SNP_cluster.append([position])\n    else:\n        SNP_cluster[-1].append(position)\nf = open(\"H2_exclude.bed\",\"w\")\nfor c in SNP_cluster:\n    #print(\"H1\", len(c), '000000F:{}-{}'.format(min(c)-50, max(c)+50), max(c)-min(c)+100)\n    print('000000F\\t{}\\t{}'.format(min(c)-50, max(c)+50), file=f)\nf.close()\nfor c in SNP_cluster:\n    print((min(c)-50, max(c)+50), max(c)-min(c)+100)", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\n./seqtk/seqtk subseq H2_pa_phase_ctg.fa H2_exclude.bed > out.fa\ngrep \">\" out.fa\nminimap2 -xasm5 MHC_37.fa  out.fa > out.paf\ncat out.paf | awk '$12 >0 {print 6\" \"$8+28477798\" \"$9+28477798\" H2:\"$1\":\"$12}' | sort -k2 -g  > H2_exclude_GRCH37.bed", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "!cat H2_exclude_GRCH37.bed", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "!cat H1_exclude_GRCH37.bed H2_exclude_GRCH37.bed | sort -k 2 -g  | tee exclude_GRCH37.bed", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "%%bash\ndx cd /20200316_asm_for_revision\ndx upload exclude_GRCH37.bed\ndx upload H2_exclude_GRCH37.bed H1_exclude_GRCH37.bed", "metadata": {"trusted": true}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}, {"cell_type": "code", "source": "", "metadata": {}, "execution_count": null, "outputs": []}]}